Add ttnncumprod multi-core implementation

jbbieniekTT opened the pull request:
### Tickets
#19892 
#21609

### Problem description
Add single-core version of `ttnn.cumprod`.

### What's changed
The existing device operation (from the scaffold, #19549 ) is to be fully implemented.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### Performance
#### Legend:
- red line signifies total composite operation max kernel time (in nanoseconds) (preprocessing + cumprod + post processing)
- green line signifies cumprod max kernel time (in nanoseconds)
- blue line signifies permute max kernel time (in nanoseconds)

#### Graphs:
- Growing acc without permutation:
![out1](https://github.com/user-attachments/assets/07417c8b-c1d0-43ed-a16c-19b7eb932749)

- Growing acc with permutation:
![out2](https://github.com/user-attachments/assets/b8b3df78-ce1a-4cfa-9081-fd3bad99541d)

- Growing work units without permutation:
![out3](https://github.com/user-attachments/assets/b204b837-dad7-4477-9c38-7093ba2ad7df)

- Growing work units with permutation:
![out4](https://github.com/user-attachments/assets/1c5fbb09-8e4a-4459-94a0-12d9e03d55af)

jbbieniekTT committed:
19582: add single-core ttnn.cumprod

- add single-core version of ttnn.cumprod implementation
that permutes (unsqueezes and rotates its input to 4D)

jbbieniekTT committed:
19892: change approach to ttnn.cumprod

Take an approach with ttnn.cumprod that allows
for any dimension of the input tensor.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp:
```diff
@@ -4,27 +4,44 @@
 namespace operations::experimental::reduction {
 
 struct CumprodOperation {
+    static constexpr uint32_t CHANNEL_DIMENSION{1};
```
Why do you need all these statics in a header?
Can they be instead declared as constants in C++?

jbbieniekTT replied:
I have removed them, I am using a different approach that doesn't require them.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp:
```diff
@@ -4,27 +4,44 @@
+
+    static std::tuple<Tensor, PermVec> permute_to_4d(const Tensor& input_tensor, const uint32_t& cum_axis);
+
+    static Tensor reorder_from_4d(
```
Can be moved out to cpp without being exposed in the header?

jbbieniekTT replied:
I have removed this, I am using a different approach that doesn't require them.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.hpp:
```diff
@@ -4,27 +4,44 @@
+        std::optional<Tensor>& optional_out,
         const std::optional<MemoryConfig>& memory_config = std::nullopt,
         const QueueId& queue_id = DefaultQueueId);
+
```
Can be moved out to cpp without being exposed in the header?

jbbieniekTT replied:
Do you mean moving the declaration to the cpp file?

jbbieniekTT committed:
19892: fix shape bug and improve pytest

- Improve test_cumprod.py to cover more cases
- Improve shape handling logic and get rid of errors
- Reformat code and address comments

jbbieniekTT committed:
19892: fix bug in the test + preal. buf.

- Fix `test_cumprod.py` - wrong dim for torch
- Fix the problem with PCC
- Fix the preallocated buffer issue

jbbieniekTT committed:
19892: PR wrap-up

just some minor cleaning

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_device_operation.hpp:
```diff
@@ -32,15 +32,22 @@ struct CumprodDeviceOperation {
+
+    struct SingleCoreCumprodProgramFactory {
+        static constexpr uint32_t CB_NUM_TILES{2};
+        enum class CumprodCB : std::underlying_type_t<tt::CBIndex> { SRC = 0, DST = 1, ONE = 16, ACC = 24 };
```
Should these be contiguous?

bbradelTT replied:
We've moved away from CB to CBIndex specifically to allow all CBs to be contiguous to increase dispatcher performance. CBs are sent in ranges, and each range is expensive. Also, we've standardized on using c_0, c_1, etc. instead of bare 0, 1, etc. although haven't updated all of the existing code.

Assuming these are the only CBs you're using and you want to keep this class, it should be something like 
```
{ SRC = tt::CBIndex::c_0, DST = tt::CBIndex::c_1, ONE = tt::CBIndex::c_2, ACC = tt::CBIndex::c_3 };
```

jbbieniekTT replied:
Understood, thank you.

jbbieniekTT replied:
Fixed.

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_device_operation.cpp:
```diff
@@ -3,55 +3,146 @@
+        tensor_args_t{input_tensor, std::move(optional_out)}};
+}
+
+hash::hash_t CumprodDeviceOperation::compute_program_hash(
```
@ayerofieiev-tt what's the current best practice for the hash?

jbbieniekTT replied:
Perhaps I could remove this? As far as I see, not a huge number of operator implementations overload this.

bbradelTT replied:
Please add a unit test enables program cache and calls the op twice with the same parameters except the tensors being separate objects..

jbbieniekTT replied:
I just included more tensor args' properties here.

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_device_operation.cpp:
```diff
@@ -3,55 +3,146 @@
+    const auto& dim{attributes.dim};
+
+    if (optional_out.has_value()) {
+        // TODO(jbbieniekTT): in this case, automatic conversion should be performed as per Torch's policy
```
Will this be done as part of this PR?
Also, should the TT_FATAL be uncommented now or when the TODO is done?

jbbieniekTT replied:
The `TT_FATAL` should be uncommented once the `dtype` attribute is handled correctly - in fact, I believe I should align first whether automatic conversion should be done at all in this case

jbbieniekTT replied:
I just realised that the conversion topic is blocked by the issue #21108, because it is required first that `ttnn.cumprod` works for integer types, making it meaningful to do any conversion here. I am going to remove this TODO for now, and I will have working conversion code in this place once integer types are supported correctly.

jbbieniekTT replied:
Removed.

bbradelTT replied:
I don't see this being removed.

jbbieniekTT replied:
Sorry, did that accidentally on some other WIP branch, please re-chek.

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/cumprod.cpp:
```diff
@@ -3,19 +3,112 @@
+    Tensor adjusted_input_tensor = input_tensor;
+    const auto& input_dtype = input_tensor.dtype();
+
+    // TODO(jbbieniekTT): enable once ttnn::to_dtype works for tensors on device storage
```
If not as part of this PR, please point to an issue.

jbbieniekTT replied:
I am aware now of the workaround which involves sending the input tensor to the CPU, processing it there and fetching it back into the device storage.

jbbieniekTT replied:
I shall remove this TODO.

jbbieniekTT replied:
Fixed.

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_writer.cpp:
```diff
@@ -0,0 +1,75 @@
+    const int32_t ACC_START_VALUE_F32{caster.u};
+    constexpr int16_t ACC_START_VALUE_F16{0x380};  // 1.0f for bfloat16
+    // TODO(jbbieniekTT): the below ones will work only if applied LLK is preconfigured appropriately for those.
+    constexpr int32_t ACC_START_VALUE_I32{0x1};
```
They will be necessary once the issue with integer type accumulation (#21108) is solved.

jbbieniekTT replied:
I am going to remove them for the time being here, and I will bring them back in the issue fix-up.

jbbieniekTT replied:
Ah, silly me. I just noticed this is the **writer** kernel, and those are completely redundant here.

jbbieniekTT replied:
Fixed.

jbbieniekTT committed:
19892: address review comments

Address review comments

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_program_factory.cpp:
```diff
@@ -3,159 +3,152 @@
+    return PLo;
+}
+
+uint32_t CumprodDeviceOperation::SingleCoreCumprodProgramFactory::calc_phi(
```
Fixed.

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_program_factory.cpp:
```diff
@@ -3,159 +3,152 @@
+    switch (input_shape.rank()) {
+        case 0: return 0;
+        case 1: return input_shape[0] / tt::constants::TILE_WIDTH;
+        default: return input_shape[input_shape.rank() - 1] * input_shape[input_shape.rank() - 2] / 1024;
```
Fixed.

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_writer.cpp:
```diff
@@ -0,0 +1,75 @@
+
+#include "dataflow_api.h"
+
+FORCE_INLINE unsigned get_tile_id(
```
Fixed.

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_writer.cpp:
```diff
@@ -0,0 +1,75 @@
+
+    for (unsigned i0 = 0; i0 < PLo; i0++) {
+        for (unsigned i1 = 0; i1 < PHi * HtWt; i1++) {
+            for (unsigned j = 0; j < tiles_per_row; j++) {
```
Fixed.

vsureshTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/compute/cumprod_single_core.cpp:
```diff
@@ -0,0 +1,91 @@
+
+            tile_regs_acquire();
+
+            mul_tiles_init(cb_in, cb_intermed);
```
This should be moved outside of the loop since the operation doesn't change between iterations. In general, the init calls are rather expensive and should be called minimal amount of times

jbbieniekTT replied:
It must be here to always initiate the appropriate pair of CBs - `cb_mul` should be here instead, and I just pushed it. Also, when I limited calling this function to at most once for every pair of CBs without repetition, it resulted in a hang.

vsureshTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_reader.cpp:
```diff
@@ -0,0 +1,106 @@
+    uint32_t tiles_per_row = get_arg_val<uint32_t>(2);  // number of tiles in a row / along axis
+    uint32_t PHi = get_arg_val<uint32_t>(3);
+    uint32_t PLo = get_arg_val<uint32_t>(4);
+    uint32_t HtWt = get_arg_val<uint32_t>(5);
```
Comments describing PHi and PLo would be a good addition

jbbieniekTT replied:
I have remade the calc util methods to calculate input_tile_offset straight up. It is a product of higher ranks, except that two last ranks need to be adjusted for tile height and width.

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_program_factory.cpp:
```diff
@@ -3,159 +3,152 @@
+uint32_t CumprodDeviceOperation::SingleCoreCumprodProgramFactory::mul_higher_ranks(
+    const Shape& input_shape, const int32_t& dim) {
+    uint32_t PHigh{1};
+    for (int32_t i{dim + 1}; i < input_shape.rank() - 2; ++i) {
```
I think the functions are calculating the. product of dimensions above and below the specified dimension.

Please
- mul_ranks_below_dim and mul_ranks_above_dim
- give better names to the variables too (prod_low and prod_high or product_low and product_high would probably be sufficient).

And prefer using `i = dim + 1` instead of `i{dim + 1}` for primitives.

Also, why not include the last dim here?

jbbieniekTT replied:
I have remade the calc util methods to calculate `input_tile_offset` straight up. It is a product of higher ranks, except that two last ranks need to be adjusted for tile height and width.

jbbieniekTT committed:
19892: add multi-core cumprod

Implement multi-core cumprod

jbbieniekTT committed:
19892: address PR comments

address PR comments

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_reader.cpp:
```diff
@@ -0,0 +1,106 @@
+                noc_async_read_barrier();
+
+                cb_push_back(cb_out, ONE_TILE);
+            }
```
Fixed.

jbbieniekTT committed:
19892: address PR comments

address PR comments

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_writer.cpp:
```diff
@@ -0,0 +1,62 @@
+            const uint32_t write_tile_id = get_tile_id(i0, i1, j, tiles_per_row, input_tile_offset);
+            cb_wait_front(cb_out, ONE_TILE);
+            uint32_t l1_read_addr = get_read_ptr(cb_out);
+            if (is_output_dram) {
```
Fixed.

jbbieniekTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_writer.cpp:
```diff
@@ -0,0 +1,62 @@
+        .bank_base_address = output_base_addr, .page_size = output_tile_bytes, .data_format = output_data_format};
+
+    for (uint32_t i = start_id; i < start_id + num_rows_per_core; ++i) {
+        const uint32_t i0 = i / input_tile_offset;
```
Provided an approach without those operations.

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/cumprod_device_operation.hpp:
```diff
@@ -53,37 +64,50 @@ struct CumprodDeviceOperation {
+            const std::vector<uint32_t>& runtime_args = {});
+
+        static uint32_t calc_input_tile_offset(const Shape& input_shape, const int32_t& dim);
+        // static uint32_t mul_nontile_ranks_above_dim(const Shape& input_shape, const int32_t& dim);
```
remove?

jbbieniekTT replied:
Done

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_reader.cpp:
```diff
@@ -0,0 +1,107 @@
+
+    const int32_t ACC_START_VALUE_F32{caster.u};
+    constexpr int32_t ACC_START_VALUE_F16{0x3F80};
+    // TODO(jbbieniekTT): the below ones will work only if applied LLK is preconfigured appropriately for those.
```
Needs to be removed or have an issue#

jbbieniekTT replied:
My bad

bbradelTT commented on ttnn/cpp/ttnn/operations/experimental/reduction/cumprod/device/kernels/dataflow/cumprod_sc_reader.cpp:
```diff
@@ -0,0 +1,107 @@
+            break;
+    }
+
+    // TODO(jbbieniekTTT): issue #
```
Fill in issue #?

jbbieniekTT replied:
My bad

jbbieniekTT committed:
19892: address PR comments

Address PR comments

jbbieniekTT committed:
19892: remove unnecessary pad filling

remove unnecessary pad filling

aczajkowskiTT commented:
Great performance graphs. I think we should push an LLK update based on this to avoid permutation.
